File: source/downloads/code/LeetCode/Java/Remove-Duplicates-from-Sorted-List2.java
public class Solution {
    public ListNode deleteDuplicates(ListNode head) {
        if(head==null || head.next==null) return head;
        ListNode newhead = new ListNode(-1);
        newhead.next = head;
        ListNode prev = newhead;
        ListNode cur = head;

        while(cur!=null){
            while(cur.next!=null && cur.next.val == cur.val)
                cur = cur.next;
            // 只有这种情况cur才会被真正加入newheadlist中
            if(prev.next == cur)
                prev = prev.next;
            else// 此时还无法判断cur.next是不是重复
                prev.next = cur.next;
            cur = cur.next;
            
        }
        return newhead.next;

    }
}

File: downloads/code/LeetCode/Python/Remove-Duplicates-from-Sorted-List2.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

 class Solution(object):
    def deleteDuplicates(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        node = ListNode(-1)
        ret = node
        while head != None:
            nextNode = head.next
            # find next node
            if nextNode != None and nextNode.val == head.val:
                while nextNode != None and nextNode.val == head.val:
                    nextNode = nextNode.next
            else:
                # found non-duplicated nodes
                node.next = head
                node = head
                # in case there is no other node
                node.next = None

            head = nextNode
        return ret.next      
